File: frontend/src/types/prediction.ts
export interface TopRunScorer {
  player_id: number
  player_name: string
  predicted_runs: number
}

export interface TopWicketTaker {
  player_id: number
  player_name: string
  predicted_wickets: number
}

export interface StartingXIPlayer {
  player_id: number
  player_name: string
  role?: string
  team_name?: string
  predicted_runs: number
  predicted_wickets: number
}

export interface ManOfTheMatch {
  player_id: number
  player_name: string
  team: string
  team_name: string
  predicted_runs: number
  predicted_wickets: number
}

export interface PredictedScores {
  home_score: number
  home_wickets: number
  away_score: number
  away_wickets: number
  first_innings_score: number
  first_innings_wickets: number
  second_innings_score: number
  second_innings_wickets: number
  first_team: string
  second_team: string
}

export interface MatchResult {
  winner: string
  result_type: 'runs' | 'wickets'
  result_text: string
  margin: number
}

export interface Top3RunScorer {
  player_id: number
  player_name: string
  predicted_runs: number
}

export interface Top3WicketTaker {
  player_id: number
  player_name: string
  predicted_wickets: number
}

export interface MatchPrediction {
  home_team: string
  away_team: string
  venue: string
  home_win_probability: number
  away_win_probability: number
  predicted_winner: string
  confidence: number
  key_factors: Array<[string, number]>
  toss_winner: string
  bat_first: string
  predicted_scores: PredictedScores
  match_result: MatchResult
  top_run_scorers: {
    home: TopRunScorer | null
    away: TopRunScorer | null
  }
  top_3_run_scorers: {
    home: Top3RunScorer[]
    away: Top3RunScorer[]
  }
  top_wicket_takers: {
    home: TopWicketTaker | null
    away: TopWicketTaker | null
  }
  top_3_wicket_takers: {
    home: Top3WicketTaker[]
    away: Top3WicketTaker[]
  }
  man_of_the_match: ManOfTheMatch | null
  predicted_starting_xi?: {
    home: StartingXIPlayer[]
    away: StartingXIPlayer[]
  }
}

export interface Standing {
  team_id: number
  avg_position: number
  avg_points: number
  position_std: number
  playoff_probability: number
  championship_probability: number
}

export interface OrangeCap {
  player_id: number
  player_name: string
  team_id?: number
  team_name?: string
  avg_runs: number
  total_runs_range: [number, number]
}

export interface PurpleCap {
  player_id: number
  player_name: string
  team_id?: number
  team_name?: string
  avg_wickets: number
  total_wickets_range: [number, number]
}

export interface Champion {
  team_id: number
  team_name?: string
  win_probability: number
}

export interface MVP {
  player_id: number
  player_name: string
  team_id?: number
  team_name?: string
  avg_runs: number
  avg_wickets: number
  mvp_score: number
}

export interface TeamOfTournamentPlayer {
  player_id: number
  player_name: string
  team_id?: number
  team_name?: string
  role: string
  avg_runs: number
  avg_wickets: number
  performance_score: number
}

export interface UpsetTracker {
  team_id: number
  team_name: string
  expected_position: number
  actual_avg_position: number
  improvement: number
}

export interface SeasonPrediction {
  predicted_standings: Standing[]
  playoff_probabilities: Record<number, number>
  championship_probabilities: Record<number, number>
  num_simulations: number
  orange_cap?: OrangeCap | null
  purple_cap?: PurpleCap | null
  champion?: Champion | null
  mvp?: MVP | null
  team_of_tournament?: TeamOfTournamentPlayer[]
  upset_tracker?: UpsetTracker[]
}

export type StandingsPrediction = Standing[]

File: backend/app/schemas/prediction.py
"""Prediction schema definitions."""
from __future__ import annotations

from typing import Dict, List, Optional

from pydantic import BaseModel


class MatchPredictionRequest(BaseModel):
    home_team_id: int
    away_team_id: int
    venue_id: int
    overwrite_venue_avg_score: Optional[float] = None
    home_lineup_player_ids: Optional[List[int]] = None
    away_lineup_player_ids: Optional[List[int]] = None


class PlayerPerformancePrediction(BaseModel):
    player_id: int
    player_name: str
    predicted_runs: Optional[float] = None
    predicted_wickets: Optional[float] = None


class PredictedScores(BaseModel):
    home_score: int
    home_wickets: int
    away_score: int
    away_wickets: int
    first_innings_score: int
    first_innings_wickets: int
    second_innings_score: int
    second_innings_wickets: int
    first_team: str
    second_team: str


class MatchResult(BaseModel):
    winner: str
    result_type: str
    result_text: str
    margin: int


class TopRunScorer(BaseModel):
    player_id: int
    player_name: str
    predicted_runs: float


class TopWicketTaker(BaseModel):
    player_id: int
    player_name: str
    predicted_wickets: float


class ManOfTheMatch(BaseModel):
    player_id: int
    player_name: str
    team: str
    team_name: str
    predicted_runs: float
    predicted_wickets: float


class Top3RunScorer(BaseModel):
    player_id: int
    player_name: str
    predicted_runs: float


class Top3WicketTaker(BaseModel):
    player_id: int
    player_name: str
    predicted_wickets: float


class StartingXIPlayer(BaseModel):
    player_id: int
    player_name: str
    role: Optional[str] = None
    team_name: Optional[str] = None
    predicted_runs: float
    predicted_wickets: float


class MatchPredictionResponse(BaseModel):
    home_team: str
    away_team: str
    venue: str
    home_win_probability: float
    away_win_probability: float
    predicted_winner: str
    confidence: float
    key_factors: List[List[str | float]]
    toss_winner: str
    bat_first: str
    predicted_scores: PredictedScores
    match_result: MatchResult
    top_run_scorers: Dict[str, Optional[TopRunScorer]]
    top_3_run_scorers: Dict[str, List[Top3RunScorer]]
    top_wicket_takers: Dict[str, Optional[TopWicketTaker]]
    top_3_wicket_takers: Dict[str, List[Top3WicketTaker]]
    man_of_the_match: Optional[ManOfTheMatch] = None
    predicted_starting_xi: Dict[str, List[StartingXIPlayer]]


class Standing(BaseModel):
    team_id: int
    avg_position: float
    avg_points: float
    position_std: float
    playoff_probability: float
    championship_probability: float


class OrangeCap(BaseModel):
    player_id: int
    player_name: str
    team_id: Optional[int] = None
    team_name: Optional[str] = None
    avg_runs: float
    total_runs_range: List[float]


class PurpleCap(BaseModel):
    player_id: int
    player_name: str
    team_id: Optional[int] = None
    team_name: Optional[str] = None
    avg_wickets: float
    total_wickets_range: List[float]


class Champion(BaseModel):
    team_id: int
    team_name: Optional[str] = None
    win_probability: float


class MVP(BaseModel):
    player_id: int
    player_name: str
    team_id: Optional[int] = None
    team_name: Optional[str] = None
    avg_runs: float
    avg_wickets: float
    mvp_score: float


class TeamOfTournamentPlayer(BaseModel):
    player_id: int
    player_name: str
    team_id: Optional[int] = None
    team_name: Optional[str] = None
    role: str
    avg_runs: float
    avg_wickets: float
    performance_score: float


class UpsetTracker(BaseModel):
    team_id: int
    team_name: str
    expected_position: float
    actual_avg_position: float
    improvement: float


class SeasonPrediction(BaseModel):
    predicted_standings: List[Standing]
    playoff_probabilities: Dict[int, float]
    championship_probabilities: Dict[int, float]
    num_simulations: int
    orange_cap: Optional[OrangeCap] = None
    purple_cap: Optional[PurpleCap] = None
    champion: Optional[Champion] = None
    mvp: Optional[MVP] = None
    team_of_tournament: List[TeamOfTournamentPlayer] = []
    upset_tracker: List[UpsetTracker] = []


class SeasonSimulationRequest(BaseModel):
    num_simulations: int = 1000
    custom_xis: Optional[Dict[int, List[int]]] = None
